jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/spat/SpatReceiverTest.java

```diff
@@ -0,0 +1,110 @@
+import us.dot.its.jpo.ode.udp.controller.UDPReceiverProperties;
+import us.dot.its.jpo.ode.util.DateTimeUtils;
+
+@RunWith(SpringRunner.class)
+@EnableConfigurationProperties
+@SpringBootTest(
+    classes = {OdeKafkaProperties.class, UDPReceiverProperties.class, KafkaProducerConfig.class},
+    properties = {
+        "ode.receivers.spat.receiver-port=15356",
+        "ode.kafka.topics.raw-encoded-json.spat=topic.SpatReceiverTest"
+    }
+)
+@ContextConfiguration(classes = {
+    UDPReceiverProperties.class,
+    RawEncodedJsonTopics.class, KafkaProperties.class
+})
+class SpatReceiverTest {
+
+  @Autowired
+  UDPReceiverProperties udpReceiverProperties;
+
+  @Autowired
+  RawEncodedJsonTopics rawEncodedJsonTopics;
+
+  @Autowired
+  private KafkaTemplate<String, String> kafkaTemplate;
+
+  EmbeddedKafkaBroker embeddedKafka = EmbeddedKafkaHolder.getEmbeddedKafka();
+
+  @Test
+  void testRun() throws Exception {
+
+    try {
+      embeddedKafka.addTopics(new NewTopic(rawEncodedJsonTopics.getSpat(), 1, (short) 1));
+    } catch (Exception e) {
+      // ignore because we only care that the topics exist not that they're unique
+    }
+
+    DateTimeUtils.setClock(
+        Clock.fixed(Instant.parse("2024-11-26T23:53:21.120Z"), ZoneId.of("UTC")));
+
+    SpatReceiver spatReceiver = new SpatReceiver(udpReceiverProperties.getSpat(), kafkaTemplate,
+        rawEncodedJsonTopics.getSpat());
+    ExecutorService executorService = Executors.newCachedThreadPool();
+    executorService.submit(spatReceiver);
+
+    String fileContent =
+        Files.readString(Paths.get(
+            "src/test/resources/us/dot/its/jpo/ode/udp/spat/SpatReceiverTest_ValidSPAT.txt"));
+
+    String expected = Files.readString(Paths.get(
+        "src/test/resources/us/dot/its/jpo/ode/udp/spat/SpatReceiverTest_ValidSPAT_expected.json"));
+
+    TestUDPClient udpClient = new TestUDPClient(udpReceiverProperties.getSpat().getReceiverPort());
+    udpClient.send(fileContent);
+
+    var consumerProps = KafkaTestUtils.consumerProps(
+        "SpatReceiverTest", "true", embeddedKafka);
+    DefaultKafkaConsumerFactory<Integer, String> cf =
+        new DefaultKafkaConsumerFactory<>(consumerProps);
+    Consumer<Integer, String> consumer = cf.createConsumer();
+    embeddedKafka.consumeFromAnEmbeddedTopic(consumer, rawEncodedJsonTopics.getSpat());
+
+    var singleRecord = KafkaTestUtils.getSingleRecord(consumer, rawEncodedJsonTopics.getSpat());
+    assertNotEquals(expected, singleRecord.value());
+    JSONObject producedJson = new JSONObject(singleRecord.value());
+    JSONObject expectedJson = new JSONObject(expected);
+
+    assertNotEquals(expectedJson.getJSONObject("metadata").get("serialId"),
+        producedJson.getJSONObject("metadata").get("serialId"));
+    expectedJson.getJSONObject("metadata").remove("serialId");
+    producedJson.getJSONObject("metadata").remove("serialId");
+
+    assertEquals(
+        expectedJson.toString(2),
```

Comment:
(non-blocking) What does passing '2' to `toString()` do for us here?

Reply:
It sets the indentation on the stringified JSON to two spaces & indicates that the JSON should be pretty-printed. Comparing pretty-printed JSON makes it easier for diff tools to identify which part of the JSON is mismatched. Without the pretty-print, the JSON is all on one line so the diff tools have a hard time highlighting specific JSON elements

Reply:
That's awesome!